### PR TITLE
[TACHYON-221] Under File System Modularization

### PR DESCRIPTION
**Pull request by calvinjia**
https://tachyon.atlassian.net/projects/TACHYON/issues/TACHYON-221

This is a resubmisison of #751 with formatting comments addressed.

The underfs modules have also been added as client dependencies as a temporary measure to make this change transparent to client jar users (a more permanent solution should be in ticket 49).

### CODE REVIEW

**Comment by haoyuan** [comment]
Thanks! Should this page be also updated ?

https://github.com/amplab/tachyon/blob/master/docs/Setup-UFS.md

**Comment by calvinjia** [comment]
Yeah, documentation pages in general haven't been updated to reflect the changes.

File: core/src/main/java/tachyon/master/TachyonMaster.java [hunk 1]
```diff
@@ -219,22 +218,14 @@ boolean isZookeeperMode() {
...
-    }
+  private void connect() throws IOException {
+    UnderFileSystem ufs = UnderFileSystem.get(mTachyonConf.get(Constants.UNDERFS_ADDRESS, null),
+        mTachyonConf);
+    ufs.connectFromMaster(mTachyonConf, NetworkUtils.getFqdnHost(mMasterAddress));
   }
 
   private void setup() throws IOException, TTransportException {
-    login();
+    connect();
```
Comment (haoyuan) [inline core/src/main/java/tachyon/master/TachyonMaster.java]: rename this to be "connectUFS"?

File: core/src/main/java/tachyon/underfs/UnderFileSystemFactory.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+   * Creates a client for accessing the given path
+   * 
+   * @param path File path
+   * @param tachyonConf Tachyon configuration
+   * @param conf Optional configuration object for the UFS, may be null
+   * @return Client
+   * @throws IllegalArgumentException Thrown if this factory does not support clients for the given
+   *         path or if the configuration provided is insufficient to create a client
+   */
+  public UnderFileSystem create(String path, TachyonConf tachyonConf, Object conf);
```
Comment (haoyuan) [inline core/src/main/java/tachyon/underfs/UnderFileSystemFactory.java]: We may want to move conf into tachyonConf. Please create a JIRA for that.

File: core/src/main/java/tachyon/underfs/UnderFileSystemFactory.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+import tachyon.conf.TachyonConf;
+
+/**
+ * Interface for under file system factories
+ */
+public interface UnderFileSystemFactory {
+  
+  /**
+   * Creates a client for accessing the given path
```
Comment (haoyuan) [inline core/src/main/java/tachyon/underfs/UnderFileSystemFactory.java]: Is the semantics "create" or "get"?

File: core/src/main/java/tachyon/underfs/UnderFileSystemFactory.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+package tachyon.underfs;
+
+import tachyon.conf.TachyonConf;
+
+/**
+ * Interface for under file system factories
+ */
+public interface UnderFileSystemFactory {
```
Comment (haoyuan) [inline core/src/main/java/tachyon/underfs/UnderFileSystemFactory.java]: Should "UnderFileSystemFactory" an interface or singleton class?

File: core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+ * <h4>Manual Registration</h4>
+ * <p>
+ * To manually register a factory simply pass an instance of your factory to the
+ * {@link #register(UnderFileSystemFactory)} method. This can be useful when your factory cannot be
+ * instantiated without arguments or in cases where automatic discovery does not give your factory
+ * priority. Factories registered this way will be registered at the start of the factories list so
+ * will have the first opportunity to indicate whether they support a requested path.
+ * </p>
+ */
+public class UnderFileSystemRegistry {
```
Comment (haoyuan) [inline core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java]: should this be a final class?

File: core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+   * @param conf Optional configuration object for the UFS, may be null
+   * @param tachyonConf Tachyon Configuration
+   * @return Client for the under file system
+   * @throws IllegalArgumentException Thrown if there is no under file system for the given path or
+   *         if no under file system could successfully be created
+   */
+  public static UnderFileSystem create(String path, TachyonConf tachyonConf, Object conf) {
+    // Try to obtain the appropriate factory
+    List<UnderFileSystemFactory> factories = findAll(path, tachyonConf);
+    if (factories.size() == 0) {
```
Comment (haoyuan) [inline core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java]: factories.empty()

File: core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+      throw new IllegalArgumentException("No known Under File System supports the given path "
```
Comment (haoyuan) [inline core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java]: The error msg is not accurate.

File: core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+   * Returns a read-only view of the available factories
+   * 
+   * @return Read-only view of the available factories
+   */
+  public static List<UnderFileSystemFactory> available() {
+    return Collections.unmodifiableList(FACTORIES);
+  }
+
+  /**
+   * Creates a client that can talk to the under file system
```
Comment (haoyuan) [inline core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java]: Creates a client of the under file system

File: core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+  }
+
+  /**
+   * Finds the first Under File System factory that supports the given path
+   * 
+   * @param path Path
+   * @param conf Tachyon configuration
+   * @return Factory if available, null otherwise
+   */
+  public static UnderFileSystemFactory find(String path, TachyonConf conf) {
```
Comment (haoyuan) [inline core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java]: why doesn't this have 'Object conf' parameter. Please create a JIRA to investigate this along with the last one.

File: core/src/main/java/tachyon/underfs/UnderFileSystemFactory.java (hunk 3 above)
Comment (calvinjia) [inline core/src/main/java/tachyon/underfs/UnderFileSystemFactory.java]: In Tachyon's factories it seems like we prefer create, but in general I think get is more widely used.

File: core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    List<UnderFileSystemFactory> eligibleFactories = new ArrayList<UnderFileSystemFactory>();
+    for (UnderFileSystemFactory factory : FACTORIES) {
+      if (factory.supportsPath(path, conf)) {
+        LOG.debug("Under File System Factory implementation {} is eligible for path {}",
+            factory.getClass(), path);
+        eligibleFactories.add(factory);
+      }
+    }
+
+    if (eligibleFactories.size() == 0) {
```
Comment (haoyuan) [inline core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java]: eligibleFactories.empty()

File: core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+    return eligibleFactories;
+  }
+
+  private static synchronized void init() {
+    if (sInit) {
+      return;
+    }
+
+    // Discover and register the available factories
+    ServiceLoader<UnderFileSystemFactory> discoverableFactories =
```
Comment (haoyuan) [inline core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java]: discoveredFactories

File: core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+   * Factories are registered at the start of the factories list so they can override the existing
+   * automatically discovered factories. Generally if you use the ServiceLoader mechanism properly
+   * it should be unnecessary to call this, however since ServiceLoader discovery order may be
+   * susceptible to class loader behavioural differences there may be rare cases when you need to
+   * manually register the desired factory.
+   * <p>
+   * 
+   * @param factory Factory to register
+   */
+  public static void register(UnderFileSystemFactory factory) {
```
Comment (haoyuan) [inline core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java]: make this "synchronized"

File: core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+    // Reinitialise
+    init();
+  }
+
+  /**
+   * Unregisters an existing factory
+   * 
+   * @param factory Factory to unregister
+   */
+  public static void unregister(UnderFileSystemFactory factory) {
```
Comment (haoyuan) [inline core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java]: synchronized

File: core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java (hunk 9 above)
Comment (calvinjia) [inline core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java]: The assumption is `supportsPath` does not require the optional ufs conf.

File: core/src/main/java/tachyon/underfs/UnderFileSystemFactory.java (hunk 2 above)
Comment (calvinjia) [inline core/src/main/java/tachyon/underfs/UnderFileSystemFactory.java]: Created TACHYON-351

File: core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java (hunk 12 above)
Comment (calvinjia) [inline core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java]: Concurrency should be handled by the CopyOnWriteArrayList?

File: core/src/main/java/tachyon/underfs/UnderFileSystem.java [hunk 14]
```diff
@@ -134,6 +137,32 @@ protected UnderFileSystem(TachyonConf tachyonConf) {
...
+   * the given master host e.g. logging in
+   * <p>
+   * Depending on the implementation this may be a no-op
+   * </p>
+   * 
+   * @param conf Tachyon configuration
+   * @param hostname The host that wants to connect to the under file system
+   * @throws IOException
+   */
+  public abstract void connectFromMaster(TachyonConf conf, String hostname) throws IOException;
```
Comment (haoyuan) [inline core/src/main/java/tachyon/underfs/UnderFileSystem.java]: It's hard to tell the difference between connectFromMaster and connectFromWorker. Please create a JIRA to investigate a better way to do / name these two metnods.

File: core/src/test/java/tachyon/client/BlockInStreamTest.java [hunk 15]
```diff
@@ -39,7 +39,9 @@
 
   @AfterClass
   public static final void afterClass() throws Exception {
-    sLocalTachyonCluster.stop();
+    if (sLocalTachyonCluster != null) {
```
Comment (haoyuan) [inline core/src/test/java/tachyon/client/BlockInStreamTest.java]: Why would this happen? Apply to the following tests.

File: core/src/test/java/tachyon/client/FileInStreamTest.java [hunk 16]
```diff
@@ -44,7 +44,11 @@
 
   @AfterClass
   public static final void afterClass() throws Exception {
-    sLocalTachyonCluster.stop();
+    if (sLocalTachyonCluster != null) {
+      sLocalTachyonCluster.stop();
+    }
+    System.clearProperty("tachyon.user.quota.unit.bytes");
+    System.clearProperty("tachyon.user.default.block.size.byte");
```
Comment (haoyuan) [inline core/src/test/java/tachyon/client/FileInStreamTest.java]: Why clear these two properties here?

File: core/src/test/java/tachyon/client/FileOutStreamTest.java [hunk 17]
```diff
@@ -68,8 +72,8 @@ public static final void beforeClass() throws IOException {
    * @param fileLen
    * @throws IOException
    */
-  private void checkWrite(TachyonURI filePath, WriteType op, int fileLen, int increasingByteArrayLen)
-      throws IOException {
+  private void checkWrite(TachyonURI filePath, WriteType op, int fileLen,
+      int increasingByteArrayLen) throws IOException {
```
Comment (haoyuan) [inline core/src/test/java/tachyon/client/FileOutStreamTest.java]: The length was more than 100 chars before?

File: core/src/test/java/tachyon/command/TFsShellTest.java [hunk 18]
```diff
@@ -80,9 +85,9 @@ public final void before() throws IOException {
 
   @Test
   public void catDirectoryTest() throws IOException {
-    String[] command = new String[] {"mkdir", "/testDir"};
+    String[] command = new String[] { "mkdir", "/testDir" };
```
Comment (haoyuan) [inline core/src/test/java/tachyon/command/TFsShellTest.java]: This is the right format?

File: core/src/test/java/tachyon/command/TFsShellTest.java [hunk 19]
```diff
@@ -130,10 +137,11 @@ public void copyFromLocalTest() throws IOException {
     File testDirInner = new File(mLocalTachyonCluster.getTachyonHome() + "/testDir/testDirInner");
     testDirInner.mkdir();
     File testFile = generateFileContent("/testDir/testFile", TestUtils.getIncreasingByteArray(10));
-    generateFileContent("/testDir/testDirInner/testFile2", TestUtils.getIncreasingByteArray(10, 20));
-    mFsShell.copyFromLocal(new String[] {"copyFromLocal", testFile.getParent(), "/testDir"});
-    Assert.assertEquals(getCommandOutput(new String[] {"copyFromLocal", testFile.getParent(),
-        "/testDir"}), mOutput.toString());
+    generateFileContent(
+        "/testDir/testDirInner/testFile2", TestUtils.getIncreasingByteArray(10, 20));
```
Comment (haoyuan) [inline core/src/test/java/tachyon/command/TFsShellTest.java]: it was more than 100 chars before?

File: core/src/test/java/tachyon/master/MasterInfoTest.java [hunk 20]
```diff
@@ -398,7 +404,8 @@ public void createFilePerfTest() throws FileAlreadyExistException, InvalidPathEx
     // System.out.println(System.currentTimeMillis() - sMs);
     // sMs = System.currentTimeMillis();
     for (int k = 0; k < 200; k ++) {
-      mMasterInfo.getClientFileInfo(new TachyonURI("/testFile").join(MasterInfo.COL + k).join("0"));
+      mMasterInfo
+          .getClientFileInfo(new TachyonURI("/testFile").join(MasterInfo.COL + k).join("0"));
```
Comment (haoyuan) [inline core/src/test/java/tachyon/master/MasterInfoTest.java]: it was more than 100 chars before?

File: core/src/test/java/tachyon/master/MasterInfoTest.java [hunk 21]
```diff
@@ -281,8 +286,8 @@ public void clientFileInfoDirectoryTest() throws InvalidPathException, FileDoesN
   }
 
   @Test
-  public void clientFileInfoEmptyFileTest() throws InvalidPathException, FileDoesNotExistException,
-      FileAlreadyExistException, BlockInfoException, TachyonException {
+  public void clientFileInfoEmptyFileTest() throws InvalidPathException,
```
Comment (haoyuan) [inline core/src/test/java/tachyon/master/MasterInfoTest.java]: it was more than 100 chars before?

File: core/src/test/java/tachyon/underfs/UnderfsUtilsTest.java [hunk 22]
```diff
@@ -62,12 +68,12 @@ public final void before() throws IOException {
 
   @Test
   public void loadUnderFsTest() throws IOException {
-    if (!UnderFileSystemCluster.isUFSHDFS()) {
-      return;
-    }
+    // TODO Is this test really tied to HDFS?
+    // Or could it run on some general subsets of Under File Systems?
+    Assume.assumeTrue(UnderFileSystemCluster.isUFSHDFS());
```
Comment (haoyuan) [inline core/src/test/java/tachyon/underfs/UnderfsUtilsTest.java]: This test is not tied to HDFS.

File: pom.xml [hunk 23]
```diff
@@ -138,12 +137,14 @@
     <license.header.path>build/license/</license.header.path>
     <checkstyle.path>build/checkstyle/</checkstyle.path>
     <findbugs.path>build/findbugs/</findbugs.path>
+    <junit.version>4.10</junit.version>
```
Comment (haoyuan) [inline pom.xml]: is this used anywhere?

File: underfs/glusterfs/pom.xml [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+      </dependencies>
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-surefire-plugin</artifactId>
+            <version>2.18.1</version>
+            <configuration>
+              <argLine>-Djava.security.krb5.realm= -Djava.security.krb5.kdc=</argLine>
+              <dependenciesToScan>
```
Comment (haoyuan) [inline underfs/glusterfs/pom.xml]: What does "dependenciesToScan" mean?

File: underfs/hdfs/src/main/java/tachyon/underfs/hdfs/HdfsUnderFileSystem.java [hunk 25]
```diff
@@ -114,6 +80,37 @@ private UnderFileSystemHdfs(String fsDefaultName, Object conf, TachyonConf tachy
...
+   * @param config Hadoop Configuration
+   */
+  protected void prepareConfiguration(String path, TachyonConf tachyonConf, Configuration config) {
+    // On Hadoop 2.x this is strictly unnecessary since it uses ServiceLoader to automatically
+    // discover available file system implementations. However this configuration setting is
+    // required for earlier Hadoop versions plus it is still honoured as an override even in 2.x so
+    // if present propagate it to the Hadoop configuration
+    String ufsHdfsImpl =
+        mTachyonConf.get(Constants.UNDERFS_HDFS_IMPL,
+            null);
```
Comment (haoyuan) [inline underfs/hdfs/src/main/java/tachyon/underfs/hdfs/HdfsUnderFileSystem.java]: format seems not right.

**Comment by haoyuan** [comment]
done a pass. thanks!

File: core/src/test/java/tachyon/client/FileOutStreamTest.java (hunk 17 above)
Comment (calvinjia) [inline core/src/test/java/tachyon/client/FileOutStreamTest.java]: Yeah not sure why checkstyle did not catch it

File: core/src/test/java/tachyon/command/TFsShellTest.java (hunk 18 above)
Comment (calvinjia) [inline core/src/test/java/tachyon/command/TFsShellTest.java]: We have mixed formatting for this case. Currently if I run it through my formatter there are no extra spaces. I can revert these changes since it doesn't seem like we have decided on a formatting method.

File: core/src/test/java/tachyon/command/TFsShellTest.java (hunk 19 above)
Comment (calvinjia) [inline core/src/test/java/tachyon/command/TFsShellTest.java]: Yeah, not sure why checkstyle did not catch it.

File: core/src/test/java/tachyon/underfs/UnderfsUtilsTest.java (hunk 22 above)
Comment (haoyuan) [inline core/src/test/java/tachyon/underfs/UnderfsUtilsTest.java]: Create another JIRA and deal it in the next PR.

File: core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java (hunk 7 above)
Comment (calvinjia) [inline core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java]: This case is when no registered ufs supports the path, whereas below one or more ufs have stated they support the path but failed to support it due to some error.

File: core/src/test/java/tachyon/worker/WorkerStorageTest.java [hunk 26]
```diff
@@ -40,6 +39,8 @@
 import tachyon.thrift.NetAddress;
 import tachyon.util.CommonUtils;
 import tachyon.worker.hierarchy.StorageDir;
+import tachyon.underfs.UnderFileSystem;
+
```
Comment (calvinjia) [inline core/src/test/java/tachyon/worker/WorkerStorageTest.java]: Missed a whitespace here, will fix with the next iteration.

File: core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java [hunk 27]
```diff
@@ -0,0 +1,256 @@
...
+   * @param tachyonConf Tachyon Configuration
+   * @return Client for the under file system
+   * @throws IllegalArgumentException Thrown if there is no under file system for the given path or
+   *         if no under file system could successfully be created
+   */
+  public static UnderFileSystem create(String path, TachyonConf tachyonConf, Object conf) {
+    // Try to obtain the appropriate factory
+    List<UnderFileSystemFactory> factories = findAll(path, tachyonConf);
+    if (factories.isEmpty()) {
+      throw new IllegalArgumentException("No known Under File System supports the given path "
```
Comment (haoyuan) [inline core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java]: IllegalArgumentException("No Under File System Factory" + path);

File: core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java (hunk 9 above)
Comment (haoyuan) [inline core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java]: Yes. Please put this issue into a JIRA issue.

File: core/src/test/java/tachyon/client/RemoteBlockInStreamTest.java [hunk 28]
```diff
@@ -70,7 +70,7 @@ public RemoteBlockInStreamTest(String[] classes) {
   public ExpectedException thrown = ExpectedException.none();
 
   @After
-  public final void after() throws Exception {
```
Comment (haoyuan) [inline core/src/test/java/tachyon/client/RemoteBlockInStreamTest.java]: an extra space here.

File: core/src/test/java/tachyon/client/TachyonFSTest.java [hunk 29]
```diff
@@ -56,10 +56,8 @@ public final void before() throws IOException {
   }
 
   @AfterClass
-  public static void afterClass() throws Exception {
+  public static final void afterClass() throws Exception {
```
Comment (haoyuan) [inline core/src/test/java/tachyon/client/TachyonFSTest.java]: why final?

File: core/src/test/java/tachyon/client/TachyonFSTest.java [hunk 30]
```diff
@@ -56,10 +56,8 @@ public final void before() throws IOException {
...
     sLocalTachyonCluster.stop();
-    System.clearProperty("tachyon.user.quota.unit.bytes");
```
Comment (haoyuan) [inline core/src/test/java/tachyon/client/TachyonFSTest.java]: interesting that this was here before.

File: core/src/test/java/tachyon/underfs/UnderFileSystemCluster.java [hunk 31]
```diff
@@ -69,18 +74,21 @@ public static UnderFileSystemCluster getUnderFilesystemCluster(String baseDir,
...
       try {
         UnderFileSystemCluster ufsCluster =
             (UnderFileSystemCluster) Class.forName(mUfsClz).getConstructor(String.class,
                 TachyonConf.class).newInstance(baseDir, tachyonConf);
+        System.out.println("Initialized under file system testing cluster of type "
+            + ufsCluster.getClass().getCanonicalName() + " for integration testing");
         return ufsCluster;
-      } catch (Exception e) {
-        System.out.println("Failed to initialize the ufsCluster of " + mUfsClz
+      } catch (Throwable e) {
```
Comment (haoyuan) [inline core/src/test/java/tachyon/underfs/UnderFileSystemCluster.java]: why this change?

File: core/src/test/java/tachyon/underfs/UnderFileSystemCluster.java [hunk 32]
```diff
@@ -96,7 +104,10 @@ public static UnderFileSystemCluster getUnderFilesystemCluster(String baseDir,
    * @return
    */
   public static boolean isUFSHDFS() {
-    return (null != mUfsClz) && (mUfsClz.equals("tachyon.LocalMiniDFSCluster"));
+    // TODO This should be renamed to something technology agnostic
+    // e.g. doZeroLengthFileReadsReturnNegative()
+    // TODO Should be dynamically determined - may need additional method on UnderFileSystem
```
Comment (haoyuan) [inline core/src/test/java/tachyon/underfs/UnderFileSystemCluster.java]: Please create a JIRA for this.

File: docs/Setup-UFS.md [hunk 33]
```diff
@@ -109,8 +109,9 @@ You should find that a tachyon folder has been created.
...
-[UnderFileSystem](https://github.com/amplab/tachyon/blob/master/core/src/main/java/tachyon/UnderFileSystem.java)
-interface. Here are the implementations of
-[HDFS](https://github.com/amplab/tachyon/blob/master/core/src/main/java/tachyon/UnderFileSystemHdfs.java)
-and [Local File System](https://github.com/amplab/tachyon/blob/master/core/src/main/java/tachyon/UnderFileSystemSingleLocal.java).
+to do so, a user should create a new underfs submodule with a
+[UnderFileSystem](https://github.com/amplab/tachyon/blob/master/core/src/main/java/tachyon/underfs/UnderFileSystem.java)
+interface and a corresponding
+[UnderFileSystemFactory](https://github.com/amplab/tachyon/blob/master/core/src/main/java/tachyon/underfs/UnderFileSystemFactory.java).
+Finally, a META-INF/services file should be included so the implementation will be registered. The
+[HDFS Submodule](https://github.com/amplab/tachyon/tree/underfs/hdfs) is an example of a complete submodule.
```
Comment (haoyuan) [inline docs/Setup-UFS.md]: Should other parts on this page remain the same after this change?

File: pom.xml [hunk 34]
```diff
@@ -121,7 +121,6 @@
   </pluginRepositories>
 
   <properties>
-    <test.profile>local</test.profile>
```
Comment (haoyuan) [inline pom.xml]: Is this needed?

Comment (haoyuan) [inline pom.xml]: If not, the doc page `contribute-to-tachyon` needs to be updated.

File: underfs/glusterfs/pom.xml [hunk 35]
```diff
@@ -0,0 +1,127 @@
...
+  <dependencies>
+    <dependency>
+      <groupId>org.tachyonproject</groupId>
+      <artifactId>tachyon-underfs-hdfs</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+
+    <!-- Test Dependencies -->
+    <dependency>
+      <groupId>junit</groupId>
```
Comment (haoyuan) [inline underfs/glusterfs/pom.xml]: doesn't `org.tachyonproject:tachyon` already contain `junit` in its test scope?

File: underfs/glusterfs/src/test/java/tachyon/underfs/glusterfs/GlusterFSCluster.java [hunk 36]
```diff
@@ -0,0 +1,65 @@
...
+
+public class GlusterFSCluster extends UnderFileSystemCluster {
+
+  public GlusterFSCluster(String baseDir, TachyonConf tachyonConf) {
+    super(baseDir, tachyonConf);
+    checkGlusterConfigured(tachyonConf);
+  }
+
+  private void checkGlusterConfigured(TachyonConf conf) {
+    if (conf == null) throw new NullPointerException("Null Tachyon Configuration provided");
```
Comment (haoyuan) [inline underfs/glusterfs/src/test/java/tachyon/underfs/glusterfs/GlusterFSCluster.java]: code style add `{}`

File: underfs/hdfs/pom.xml [hunk 37]
```diff
@@ -0,0 +1,121 @@
...
+    <!-- Test Dependencies -->
+    <dependency>
+      <groupId>org.tachyonproject</groupId>
+      <artifactId>tachyon</artifactId>
+      <version>${project.version}</version>
+      <classifier>tests</classifier>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.thrift</groupId>
```
Comment (haoyuan) [inline underfs/hdfs/pom.xml]: why adding thrift here?

File: core/src/test/java/tachyon/client/TachyonFSTest.java (hunk 29 above)
Comment (calvinjia) [inline core/src/test/java/tachyon/client/TachyonFSTest.java]: Seems like most of our `afterClass()` are `public static final` I don't see any reason why they can't just be `public static`.

File: underfs/hdfs/src/main/java/tachyon/underfs/hdfs/HdfsUnderFileSystem.java [hunk 38]
```diff
@@ -305,8 +300,32 @@ public boolean isFile(String path) throws IOException {
...
-      String hostname) throws IOException {
+  @Override
+  public void connectFromMaster(TachyonConf conf, String host) throws IOException {
+    String masterKeytab = conf.get(Constants.MASTER_KEYTAB_KEY, null);
+    String masterPrincipal = conf.get(Constants.MASTER_PRINCIPAL_KEY, null);
+    if (masterKeytab == null || masterPrincipal == null) {
+      return;
+    }
+
+    this.login(Constants.MASTER_KEYTAB_KEY, masterKeytab, Constants.MASTER_PRINCIPAL_KEY,
```
Comment (haoyuan) [inline underfs/hdfs/src/main/java/tachyon/underfs/hdfs/HdfsUnderFileSystem.java]: remove `this.`

File: underfs/hdfs/src/main/java/tachyon/underfs/hdfs/HdfsUnderFileSystem.java [hunk 39]
```diff
@@ -305,8 +300,32 @@ public boolean isFile(String path) throws IOException {
...
+
+  @Override
+  public void connectFromWorker(TachyonConf conf, String host) throws IOException {
+    String workerKeytab = conf.get(Constants.WORKER_KEYTAB_KEY, null);
+    String workerPrincipal = conf.get(Constants.WORKER_PRINCIPAL_KEY, null);
+    if (workerKeytab == null || workerPrincipal == null) {
+      return;
+    }
+
+    this.login(Constants.WORKER_KEYTAB_KEY, workerKeytab, Constants.WORKER_PRINCIPAL_KEY,
```
Comment (haoyuan) [inline underfs/hdfs/src/main/java/tachyon/underfs/hdfs/HdfsUnderFileSystem.java]: remove `this.`

File: underfs/hdfs/pom.xml [hunk 40]
```diff
@@ -0,0 +1,121 @@
...
+    <version>0.7.0-SNAPSHOT</version>
+  </parent>
+  <artifactId>tachyon-underfs-hdfs</artifactId>
+  <name>Tachyon Under File System - HDFS</name>
+  <description>HDFS Under File System implementation</description>
+
+  <properties>
+    <license.header.path>${project.parent.parent.basedir}/build/license/</license.header.path>
+    <checkstyle.path>${project.parent.parent.basedir}/build/checkstyle/</checkstyle.path>
+    <findbugs.path>${project.parent.parent.basedir}/build/findbugs/</findbugs.path>
```
Comment (haoyuan) [inline underfs/hdfs/pom.xml]: does pom file inherit properties values from its parent pom files? if yes, these values has been set in its parent pom files. 

If this requires verification, please create a jira and do it in another pr.

**Comment by haoyuan** [comment]
done a pass.

File: core/src/main/java/tachyon/underfs/UnderFileSystem.java (hunk 14 above)
Comment (calvinjia) [inline core/src/main/java/tachyon/underfs/UnderFileSystem.java]: https://tachyon.atlassian.net/browse/TACHYON-352

File: core/src/test/java/tachyon/underfs/UnderfsUtilsTest.java (hunk 22 above)
Comment (calvinjia) [inline core/src/test/java/tachyon/underfs/UnderfsUtilsTest.java]: https://tachyon.atlassian.net/browse/TACHYON-353

File: core/src/test/java/tachyon/underfs/UnderFileSystemCluster.java (hunk 32 above)
Comment (calvinjia) [inline core/src/test/java/tachyon/underfs/UnderFileSystemCluster.java]: https://tachyon.atlassian.net/browse/TACHYON-354

File: underfs/hdfs/pom.xml (hunk 40 above)
Comment (calvinjia) [inline underfs/hdfs/pom.xml]: https://tachyon.atlassian.net/browse/TACHYON-355 (for all the other pom related comments as well)

File: pom.xml (hunk 34 above)
Comment (calvinjia) [inline pom.xml]: https://tachyon.atlassian.net/browse/TACHYON-356

File: core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java (hunk 9 above)
Comment (calvinjia) [inline core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java]: Should also go into TACHYON-351

File: underfs/hdfs/pom.xml (hunk 37 above)
Comment (calvinjia) [inline underfs/hdfs/pom.xml]: https://tachyon.atlassian.net/browse/TACHYON-355

File: underfs/glusterfs/pom.xml (hunk 35 above)
Comment (calvinjia) [inline underfs/glusterfs/pom.xml]: https://tachyon.atlassian.net/browse/TACHYON-355

File: core/src/main/java/tachyon/worker/TachyonWorker.java [hunk 41]
```diff
@@ -243,19 +242,10 @@ WorkerServiceHandler getWorkerServiceHandler() {
     return mWorkerServiceHandler;
   }
 
-  private void login() throws IOException {
-    String workerKeytabFile = mTachyonConf.get(Constants.WORKER_KEYTAB_KEY, null);
-    String workerPrincipal = mTachyonConf.get(Constants.WORKER_PRINCIPAL_KEY, null);
-    if (workerKeytabFile == null || workerPrincipal == null) {
-      return;
-    }
+  private void connect() throws IOException {
```
Comment (calvinjia) [inline core/src/main/java/tachyon/worker/TachyonWorker.java]: I'll rename this to connectToUFS in the next iteration to keep it consistent with that we did on the master side.

File: underfs/hdfs/pom.xml (hunk 40 above)
Comment (rvesse) [inline underfs/hdfs/pom.xml]: These paths need to be set if you want to be able to build individual modules

A top level build will inherit the top level properties but if these are not set a single module build will fail because the inherited properties have incorrect values for a single module build
